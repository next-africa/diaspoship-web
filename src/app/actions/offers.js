import { createAction } from 'redux-actions';
import offers from '../../data/offers.json';

export const selectOffer = createAction('SELECT_OFFER');
export const receiveOffers = createAction('RECEIVE_OFFERS');
export const fetchingOffers = createAction('FETCHING_OFFERS');

const fetchOffersFn = dispatch => {
  dispatch(fetchingOffers());
  dispatch(receiveOffers(offers));
};

export const fetchOffers = () => fetchOffersFn;
