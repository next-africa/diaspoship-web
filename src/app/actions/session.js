import { createAction } from 'redux-actions';
import { FB_APP_ID } from '../../config/environment';
export const resetUser = createAction('LOGOUT_USER');
export const isUserConnected = createAction('IS_USER_CONNECTED');
export const userIsConnecting = createAction('USER_IS_CONNECTING');
export const setUser = createAction('USER_FETCH_DATA_SUCCESS');
const FB_APP_ID = process.env.REACT_APP_SECRET_FB_APP_ID;

export function login() {
  return dispatch => {
    dispatch(userIsConnecting(true));
    window.FB.login(function(response) {
      if (response && !response.error) {
        dispatch(loadUserInfos());
      } else {
        dispatch(userIsConnecting(false));
      }
    });
  };
}

export function logout() {
  return dispatch => {
    window.FB.logout();
    dispatch(resetUser());
  };
}

export function loadUserInfos() {
  return dispatch => {
    window.FB.api('/me', { fields: 'id, name, email, picture' }, function(
      response
    ) {
      if (response && !response.error) {
        const { id, email, name, picture } = response;
        dispatch(
          setUser({
            id,
            email,
            name: name.split(' ')[0],
            picture: picture.data.url
          })
        );
      }
      dispatch(userIsConnecting(false));
    });
  };
}

export function updateUserLoginStatus() {
  return dispatch => {
    dispatch(userIsConnecting(true));
    window.FB.getLoginStatus(function(response) {
      if (response && response.status === 'connected') {
        dispatch(loadUserInfos());
      } else {
        dispatch(userIsConnecting(false));
        dispatch(resetUser());
      }
    });
  };
}

export function initializeFacebookSDK() {
  return dispatch => {
    window.fbAsyncInit = function() {
      window.FB.init({
        appId: FB_APP_ID,
        cookie: true,
        autoLoad: true,
        xfbml: true,
        version: 'v2.10'
      });

      dispatch(updateUserLoginStatus());
    };

    // Load the SDK asynchronously
    (function(d, s, id) {
      let js,
        fjs = d.getElementsByTagName(s)[0];
      if (d.getElementById(id) || !fjs) return;
      js = d.createElement(s);
      js.id = id;
      js.src = '//connect.facebook.net/en_US/sdk.js';
      fjs.parentNode.insertBefore(js, fjs);
    })(document, 'script', 'facebook-jssdk');
  };
}
