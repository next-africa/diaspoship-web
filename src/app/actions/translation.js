// Redux
import { createAction } from 'redux-actions';

// Cookies
import Cookies from 'universal-cookie';

const cookies = new Cookies();

export const setLanguage = createAction('SET_LANGUAGE');

export const selectLanguage = locale => {
  return dispatch => {
    if (!locale) locale = cookies.get('locale');

    if (!locale) locale = window.navigator.language.split('-')[0];

    cookies.set('locale', locale);
    return dispatch(setLanguage(locale));
  };
};
