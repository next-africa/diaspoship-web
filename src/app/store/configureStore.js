import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import diasposhipApp from '../reducers';

export default function configureStore(initialState) {
  return createStore(diasposhipApp, initialState, applyMiddleware(thunk));
}
