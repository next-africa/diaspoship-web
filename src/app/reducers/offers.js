// Redux
import { handleActions } from 'redux-actions';

import { getOffer } from '../actions/offers';

export const INITIAL_STATE = {
  offersList: [
    {
      id: 1,
      locationFrom: 'Quebec',
      locationTo: 'Montreal',
      kilo: '5 kg',
      address: '2325 rue de la vie etudiante ',
      date: '20170204'
    },
    {
      id: 3,
      locationFrom: 'Senegal',
      locationTo: 'Kinshasa',
      kilo: '1 kg',
      address: 'Pikine SEFA ',
      date: '20170209'
    },
    {
      id: 4,
      locationFrom: 'Paris',
      locationTo: 'Montreal',
      kilo: '3 kg',
      address: '2325 rue de la vie etudiante ',
      date: '20170204'
    }
  ],
  selectedOffer: null
};

export default handleActions(
  {
    [getOffer]: (state, { payload }) => {
      let offer = state.offersList.filter(offer => {
        if (offer.id === payload) {
          return offer;
        }
        return 0;
      });
      return {
        ...state,
        selectedOffer: offer[0]
      };
    }
  },
  INITIAL_STATE
);
