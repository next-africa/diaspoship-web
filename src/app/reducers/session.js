import { handleActions } from 'redux-actions';
import { logoutUser, loginUser } from '../actions/session';

const INITIAL_STATE = {
  connected: false,
  userInfos: undefined
};
export default handleActions(
  {
    [loginUser]: (state, { payload }) => {
      let connected = state.connected;
      if (payload.profile.status === 'unknown' || connected) {
        return state;
      }
      localStorage.setItem('isConnected', true);
      return Object.assign({}, state, {
        connected: true,
        userInfos: payload.profile
      });
    },
    [logoutUser]: (state, { payload }) => {
      localStorage.setItem('isConnected', false);
      return Object.assign({}, state, {
        connected: false,
        userInfos: undefined
      });
    }
  },
  INITIAL_STATE
);
