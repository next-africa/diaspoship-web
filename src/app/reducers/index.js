import { combineReducers } from 'redux';
import translation from './translation';
import session from './session';
import offers from './offers';

export default combineReducers({
  session,
  translation,
  offers
});
