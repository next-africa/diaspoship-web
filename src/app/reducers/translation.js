import { handleActions } from 'redux-actions';
import { DEFAULT_LANGUAGE, AVAILABLE_LANGUAGES } from '../constants';

import { setLanguage } from '../actions/translation';

const INITIAL_STATE = {
  selectedLanguage: DEFAULT_LANGUAGE,
  selectedTranslations: AVAILABLE_LANGUAGES[DEFAULT_LANGUAGE]
};

export default handleActions(
  {
    [setLanguage]: (state, { payload }) => ({
      selectedLanguage: payload,
      selectedTranslations: AVAILABLE_LANGUAGES[payload]
    })
  },
  INITIAL_STATE
);
