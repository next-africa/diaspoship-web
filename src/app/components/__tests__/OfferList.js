import React from 'react';
import ReactDOM from 'react-dom';
import configureStore from 'redux-mock-store';
import { mount } from 'enzyme';
import '../../../test-util/enzyme-configuration';
import createComponentWithIntl from '../../../test-util/create-component-with-intl';
import renderer from 'react-test-renderer';
import createConnectedComponent from '../../../test-util/create-connected-component';
import thunk from 'redux-thunk';

import { INITIAL_STATE } from '../../reducers/offers';
import OfferListContainer, { OfferList } from '../OfferList';
import { fetchingOffers } from '../../actions/offers';

describe('OfferList component', () => {
  let fetchOffersMock;

  beforeEach(() => {
    fetchOffersMock = jest.fn();
  });

  it('does not crash', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      createComponentWithIntl(
        <OfferList
          offers={[]}
          isFetching={false}
          onFetchOffers={fetchOffersMock}
        />
      ),
      div
    );
  });

  it('renders correctly when offers are being fetched', () => {
    const tree = renderer
      .create(
        createComponentWithIntl(
          <OfferList
            offers={[]}
            isFetching={true}
            onFetchOffers={fetchOffersMock}
          />
        )
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when offers are not being fetched', () => {
    const tree = renderer
      .create(
        createComponentWithIntl(
          <OfferList
            offers={[]}
            isFetching={false}
            onFetchOffers={fetchOffersMock}
          />
        )
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('fetches offers when it is first rendered and it is not already fetching', () => {
    mount(
      createComponentWithIntl(
        <OfferList
          offers={[]}
          isFetching={false}
          onFetchOffers={fetchOffersMock}
        />
      )
    );

    expect(fetchOffersMock.mock.calls.length).toBe(1);
  });

  it('should not fetch offers when it is first rendered and it is already fetching', () => {
    mount(
      createComponentWithIntl(
        <OfferList
          offers={[]}
          isFetching={true}
          onFetchOffers={fetchOffersMock}
        />
      )
    );

    expect(fetchOffersMock.mock.calls.length).toBe(0);
  });
});

describe('OfferList Container', () => {
  const mockStore = configureStore([thunk])({
    offers: INITIAL_STATE
  });

  const root = mount(
    createConnectedComponent(createComponentWithIntl(<OfferListContainer />), {
      store: mockStore
    })
  );

  const offerList = root.find(OfferList);

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      createConnectedComponent(
        createComponentWithIntl(<OfferListContainer />),
        { store: mockStore }
      ),
      div
    );
  });

  it('is connected to the store', () => {
    expect(offerList.prop('error')).toEqual(INITIAL_STATE.error);
    expect(offerList.prop('isFetching')).toEqual(INITIAL_STATE.isFetching);
    expect(offerList.prop('offers')).toEqual(INITIAL_STATE.offers);
    expect(offerList.prop('filters')).toEqual(INITIAL_STATE.filters);
  });

  it('dispatches fetchOffers when onFetchOffers is called', () => {
    offerList.prop('onFetchOffers')();

    expect(mockStore.getActions()[0]).toEqual(fetchingOffers());
  });
});
