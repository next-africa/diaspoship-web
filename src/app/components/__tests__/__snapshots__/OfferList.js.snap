// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OfferList component renders correctly when offers are being fetched 1`] = `
<div
  className="OfferListComponent-root-1"
>
  <div
    className="OfferListComponent-offerListHeader-3"
  >
    <h4>
      <span>
        Latest offers
      </span>
    </h4>
    <div
      className="MuiLinearProgress-root-5 MuiLinearProgress-primaryColor-6 MuiLinearProgress-rootQuery-16"
    >
      <div
        className="MuiLinearProgress-bar-12 MuiLinearProgress-primaryColorBar-7 MuiLinearProgress-indeterminateBar1-17"
        style={Object {}}
      />
      <div
        className="MuiLinearProgress-bar-12 MuiLinearProgress-primaryColorBar-7 MuiLinearProgress-indeterminateBar2-18"
        style={Object {}}
      />
    </div>
  </div>
  <div
    className="OfferListComponent-offerListContent-2"
  />
</div>
`;

exports[`OfferList component renders correctly when offers are not being fetched 1`] = `
<div
  className="OfferListComponent-root-1"
>
  <div
    className="OfferListComponent-offerListHeader-3"
  >
    <h4>
      <span>
        Latest offers
      </span>
    </h4>
  </div>
  <div
    className="OfferListComponent-offerListContent-2"
  />
</div>
`;
